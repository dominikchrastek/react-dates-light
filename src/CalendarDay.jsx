/* @flow */
import * as React from "react";
import styled from "styled-components";
import type { ButtonProps, CalendarDayProps } from ".";

type Props = CalendarDayProps;

// export const getClasses = (props: Props) => {
//   if (props.isPast || props.isFuture || props.isDisabled) {
//     return props.classes.disabled;
//   }
//   if (props.isHovered) {
//     return props.classes.selected;
//   }
//   if (props.isSelected) {
//     return props.classes.selected;
//   }

//   return props.classes.date;
// };

export const buttonColor = (props: ButtonProps) => {
  if (props.className) {
    return "inherit";
  }
  if (props.disabled) {
    return props.colors.disabled;
  }
  if (props.isHovered) {
    return "white";
  }
  if (props.isSelected) {
    return "white";
  }

  return "inherit";
};

export const buttonBg = (props: ButtonProps) => {
  if (props.className) {
    return "inherit";
  }
  if (props.disabled) {
    return props.colors.background;
  }

  if (props.isHovered) {
    return props.colors.selected;
  }
  if (props.isSelected) {
    return props.colors.selected;
  }

  return "white";
};

export const getHover = (props: ButtonProps) => {
  if (props.isHovered || props.isFocused) {
    return props.colors.selected;
  }
  if (props.isSelected) {
    return props.colors.selectedHover;
  }
  return props.colors.hover;
};

export const getHoverColor = (props: ButtonProps) => {
  if (props.disabled) {
    return props.colors.disabled;
  }
  if (props.isFocused || props.isSelected) {
    return "white";
  }
  return "inherit";
};

const Button = styled.button`
  width: 100%;
  height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
  opacity: ${props => (props.isHidden ? 0 : 1)};
  color: ${props => buttonColor(props)};
  background: ${props => buttonBg(props)};
  border: none;
  font-size: inherit;

  ${props => !props.isHidden && "cursor: pointer;"};
  border-radius: 0;
  :hover {
    color: ${props => getHoverColor(props)};
    background: ${props => getHover(props)};
  }
  :disabled {
    cursor: default;
    :hover {
      background: inherit;
    }
  }
  :focus {
    outline: none;
  }
`;

const Td = styled.div`
  display: table-cell;
  border: ${props =>
    props.isHidden ? "none" : `1px solid ${props.colors.border}`};
  box-sizing: border-box;
`;

export default class CalendarDay extends React.PureComponent<Props> {
  handleClick = () => {
    const { value, isHidden, selectDate } = this.props;
    if (!isHidden) {
      selectDate(value);
    }
  };

  handleHover = () => {
    const { value, isHidden, onHover } = this.props;
    if (!isHidden) {
      onHover(value);
    }
  };

  props: Props;

  render() {
    const {
      isHidden,
      number,
      isPast,
      isFuture,
      isSelected,
      isDisabled,
      isHovered,
      colors,
      isFocused,
      customClasses,
      classes
    } = this.props;

    if (classes.day) {
      customClasses.push(classes.day);
    }

    return (
      <Td
        isHidden={isHidden}
        colors={colors}
        className={
          (customClasses &&
            (!isHovered && !isSelected) &&
            customClasses.join(" ")) ||
          classes.day
        }
      >
        <Button
          // className={getClasses(this.props)}
          onClick={this.handleClick}
          onMouseOver={this.handleHover}
          onFocus={this.handleHover}
          disabled={isPast || isFuture || isDisabled}
          isSelected={isSelected}
          isHidden={isHidden}
          isPast={isPast}
          isFuture={isFuture}
          isHovered={isHovered}
          isDisabled={isDisabled}
          colors={colors}
          isFocused={isFocused}
          number={number}
        >
          {number}
        </Button>
      </Td>
    );
  }
}
