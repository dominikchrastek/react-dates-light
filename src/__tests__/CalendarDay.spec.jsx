import React from "react";
import { shallow } from "enzyme";

import getDate from "date-fns/getDate";

import CalendarDay, {
  buttonBg,
  buttonColor,
  getHover,
  getHoverColor
} from "../CalendarDay";

describe("#CalendarDay", () => {
  const date = new Date(2018, 1, 1);
  const classes = {
    disabled: "disabled",
    selected: "selected",
    date: "default"
  };

  const colors = {
    inherit: "inherit",
    white: "white",
    selected: "selected",
    disabled: "disabled",
    background: "background",
    selectedHover: "selectedHover",
    hover: "hover"
  };
  describe("methods", () => {
    it("should render correctly", () => {
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
    it("should render with class day", () => {
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{
            day: "day"
          }}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
    it("should render joined classes", () => {
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={["hey", "hou"]}
          classes={{}}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
    it("should render with joined & class day", () => {
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={["hey", "hou"]}
          classes={{
            day: "day"
          }}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });
    it("should render hidden", () => {
      const wrapper = shallow(
        <CalendarDay
          isHidden
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it("should handleHover", () => {
      const onHover = jest.fn();
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={onHover}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );

      const node = wrapper.instance();
      node.handleHover();

      expect(onHover).toBeCalledWith(date);
    });

    it("should not handleHover (isHidden)", () => {
      const onHover = jest.fn();
      const wrapper = shallow(
        <CalendarDay
          isHidden
          number={getDate(date)}
          value={date}
          selectDate={jest.fn()}
          onHover={onHover}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );

      const node = wrapper.instance();
      node.handleHover();

      expect(onHover).not.toBeCalled();
    });

    it("should handleClick", () => {
      const onClick = jest.fn();
      const wrapper = shallow(
        <CalendarDay
          isHidden={false}
          number={getDate(date)}
          value={date}
          selectDate={onClick}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );

      const node = wrapper.instance();
      node.handleClick();

      expect(onClick).toBeCalledWith(date);
    });

    it("should not handleClick (isHidden)", () => {
      const onClick = jest.fn();
      const wrapper = shallow(
        <CalendarDay
          isHidden
          number={getDate(date)}
          value={date}
          selectDate={onClick}
          onHover={jest.fn()}
          isHovered={false}
          isSelected={false}
          isPast={false}
          isFuture={false}
          colors={{}}
          customClasses={[]}
          classes={{}}
        />
      );

      const node = wrapper.instance();
      node.handleClick();

      expect(onClick).not.toBeCalled();
    });
  });

  describe("helpers", () => {
    const props = (prop, className = null) => ({
      [prop]: true,
      className,
      classes,
      colors
    });
    // it('getClasses', () => {
    //   expect(getClasses(props('isPast'))).toBe(classes.disabled);
    //   expect(getClasses(props('isFuture'))).toBe(classes.disabled);
    //   expect(getClasses(props('isDisabled'))).toBe(classes.disabled);
    //   expect(getClasses(props('isHovered'))).toBe(classes.selected);
    //   expect(getClasses(props('isSelected'))).toBe(classes.selected);
    //   expect(getClasses({ classes })).toBe(classes.date);
    // });
    it("buttonColor", () => {
      expect(buttonColor(props("isPast", "class"))).toBe(colors.inherit);
      expect(buttonColor(props("isFuture"))).toBe(colors.inherit);
      expect(buttonColor(props("isPast"))).toBe(colors.inherit);
      expect(buttonColor(props("disabled"))).toBe(colors.disabled);
      expect(buttonColor(props("isHovered"))).toBe(colors.white);
      expect(buttonColor(props("isSelected"))).toBe(colors.white);
      expect(buttonColor({ colors })).toBe(colors.inherit);
    });
    it("buttonBg", () => {
      expect(buttonBg(props("isPast", "class"))).toBe(colors.inherit);
      expect(buttonBg(props("isFuture"))).toBe(colors.white);
      expect(buttonBg(props("isPast"))).toBe(colors.white);
      expect(buttonBg(props("disabled"))).toBe(colors.background);
      expect(buttonBg(props("isHovered"))).toBe(colors.selected);
      expect(buttonBg(props("isSelected"))).toBe(colors.selected);
      expect(buttonBg({ colors })).toBe(colors.white);
    });
    it("getHover", () => {
      expect(getHover(props("isHovered"))).toBe(colors.selected);
      expect(getHover(props("isFocused"))).toBe(colors.selected);
      expect(getHover(props("isSelected"))).toBe(colors.selectedHover);
      expect(getHover(props("isPast"))).toBe(colors.hover);
    });
    it("getHoverColor", () => {
      expect(getHoverColor(props("disabled"))).toBe(colors.disabled);
      expect(getHoverColor(props("isFocused"))).toBe(colors.white);
      expect(getHoverColor(props("isSelected"))).toBe(colors.white);
      expect(getHoverColor(props("isPast"))).toBe(colors.inherit);
    });
  });
});
