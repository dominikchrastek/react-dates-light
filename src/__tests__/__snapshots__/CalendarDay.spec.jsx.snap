// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CalendarDay methods should render correctly 1`] = `
<ForwardRef
  colors={Object {}}
  isHidden={false}
>
  <ForwardRef
    colors={Object {}}
    isFuture={false}
    isHidden={false}
    isHovered={false}
    isPast={false}
    isSelected={false}
    number={1}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </ForwardRef>
</ForwardRef>
`;

exports[`#CalendarDay methods should render hidden 1`] = `
<ForwardRef
  colors={Object {}}
  isHidden={true}
>
  <ForwardRef
    colors={Object {}}
    isFuture={false}
    isHidden={true}
    isHovered={false}
    isPast={false}
    isSelected={false}
    number={1}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </ForwardRef>
</ForwardRef>
`;

exports[`#CalendarDay methods should render joined classes 1`] = `
<ForwardRef
  className="hey hou"
  colors={Object {}}
  isHidden={false}
>
  <ForwardRef
    colors={Object {}}
    isFuture={false}
    isHidden={false}
    isHovered={false}
    isPast={false}
    isSelected={false}
    number={1}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </ForwardRef>
</ForwardRef>
`;

exports[`#CalendarDay methods should render with class day 1`] = `
<ForwardRef
  className="day"
  colors={Object {}}
  isHidden={false}
>
  <ForwardRef
    colors={Object {}}
    isFuture={false}
    isHidden={false}
    isHovered={false}
    isPast={false}
    isSelected={false}
    number={1}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </ForwardRef>
</ForwardRef>
`;

exports[`#CalendarDay methods should render with joined & class day 1`] = `
<ForwardRef
  className="hey hou day"
  colors={Object {}}
  isHidden={false}
>
  <ForwardRef
    colors={Object {}}
    isFuture={false}
    isHidden={false}
    isHovered={false}
    isPast={false}
    isSelected={false}
    number={1}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
  >
    1
  </ForwardRef>
</ForwardRef>
`;
